File: leetcode/medium/78. Subsets/solution.js
// https://leetcode.com/problems/subsets/
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
function subsets(nums) {
    const powerset = [];
    const makePowerSet = (path, index) => {
        powerset.push(path);
        for (let i = index; i < nums.length; i += 1) {
            makePowerSet([...path, nums[i]], i + 1);
        }
    };

    makePowerSet([], 0);
    return powerset;
};

subsets([1, 2, 3])

File: leetcode/medium/78. Subsets/solution.cc
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> path;
        recursion(res, nums, path, 0);
        return res;
    }

    void recursion(vector<vector<int>>& res, vector<int>& nums, vector<int>& path, int index) {
        res.push_back(path);
        for (int i = index; i < nums.size(); i += 1) {
            path.push_back(nums[i]);
            recursion(res, nums, path, i + 1);
            path.pop_back();
        }
        return;
    }
};